content/blog/2019-12-18-halide-autoscheduler/index.md

```diff
@@ -0,0 +1,359 @@
+The images below are from that document. 
+
+[jrkthesis]: https://people.csail.mit.edu/jrk/jrkthesis.pdf
+
+In order to search for schedule, we represent them as *schedule trees*,
+wherein the ancestry relationships between nodes represent ordering information.
+Schedule trees have the following kinds of nodes:
+
+- **Root nodes** represent the top of the schedule tree.
+
+- **Loop nodes** represent the traversal of how the function is computed
+  along a given dimension.
+  Loop nodes are associated with a function and a variable (dimension).
+  Since functions are assumed two-dimensional, by default functions have
+  two variables: *x* and *y*.
+  Loop nodes also contain information such as whether the loop is run
+  sequentially, run in parallel, or vectorized.
+
+- **Storage nodes** represent storage for intermediate results to be used later.
+
+- **Compute nodes** are the leaves of the schedule tree, and they represent
+  computation being performed.
+  Compute nodes can have other compute nodes as children to represent
+  functions that are inlined instead of loaded from intermediate storage.
+
+Schedule trees are considered *well-formed* if they satisfy the
+following criteria:
+
+- The ancestry path from a function's compute node to the root node contains
+  all the loop nodes and the storage node
+  (if the function is not the output[^storage-output])
+  for that function.
+  Intuitively, this means that the traversal of how the function is computed
+  is completely defined, and storage for the function's results is available.
+ 
+- If a function calls another function and the callee is not inlined,
+  the compute node for the callee must occur before the compute node of the
+  caller in a depth-first traversal.
+  Intuitively, this ensures that the callee's results are stored before
+  the caller is computed.
+
+[^storage-output]: By convention the output function does not have a storage
+  node in the schedule tree since it is assumed that storage for the output
+  has already been allocated and thus there is no decision to be made about
+  the granularity with which to allocate it.
+
+For any function we can define the *default schedule*, which traverses
+the output function in row-major order and inlines all called functions,
+like so:
+
+![default schedule](default.png)
+
+We can give a semantics for schedule trees as nested loops.
+Consider the schedule below for three functions, *in*, *bx*, and *by*, where
+*by* calls *bx* and *bx* calls *in*.
+The schedule tree on the left represents the nested loop on the right.
+
+![schedule semantics](semantics.png)
+
+
+### Schedule Transformers
+
+We define transformers over schedule trees.
+We use these to traverse the search space of schedules.
+
+- **Split** - split a function's variable into two. For example, we can split
+  a function's `x` variable into `x_inner` and `x_outer`.
+  Combined with *reorder*, *split* can represent schedules that *tile*
+  computations.
```

Comment:
It makes sense that something like this would be useful for tiling, but the description is still a bit vague—maybe add one more half-sentence to explain what `x_inner` and `x_outer` would represent in this example?